playground/backend/cmd/server/controller.go

```diff
@@ -203,55 +212,154 @@ func setupValidators(sdk pb.Sdk, filepath string) *[]validators.Validator {
 
 // processCode validates, compiles and runs code by pipelineId.
 // During each operation updates status of execution and saves it into cache:
+// - In case of processing works more that timeout duration saves playground.Status_STATUS_RUN_TIMEOUT as cache.Status into cache.
+// - In case of code processing has been canceled saves playground.Status_STATUS_CANCELED as cache.Status into cache.
 // - In case of validation step is failed saves playground.Status_STATUS_ERROR as cache.Status into cache.
 // - In case of compile step is failed saves playground.Status_STATUS_COMPILE_ERROR as cache.Status and compile logs as cache.CompileOutput into cache.
 // - In case of compile step is completed with no errors saves empty string ("") as cache.CompileOutput into cache.
 // - In case of run step is failed saves playground.Status_STATUS_ERROR as cache.Status and run logs as cache.RunOutput into cache.
 // - In case of run step is completed with no errors saves playground.Status_STATUS_FINISHED as cache.Status and run output as cache.RunOutput into cache.
 // At the end of this method deletes all created folders.
 func processCode(ctx context.Context, cacheService cache.Cache, lc *fs_tool.LifeCycle, compileBuilder *executors.CompileBuilder, pipelineId uuid.UUID, env *environment.Environment, sdk pb.Sdk) {
-	defer cleanUp(pipelineId, lc)
+	ctxWithTimeout, cancelByTimeoutFunc := context.WithTimeout(ctx, env.ApplicationEnvs.PipelineExecuteTimeout())
+	defer func(lc *fs_tool.LifeCycle) {
+		cancelByTimeoutFunc()
+		cleanUp(pipelineId, lc)
+	}(lc)
+
+	errorChannel := make(chan error, 1)
+	dataChannel := make(chan interface{}, 1)
+	doneChannel := make(chan bool, 1)
+	cancelChan := make(chan bool, 1)
+
+	go cancelCheck(ctx, pipelineId, cancelChan, cacheService)
 
 	// build executor for validate and compile steps
 	exec := compileBuilder.Build()
 
 	// validate
 	logger.Infof("%s: Validate() ...\n", pipelineId)
 	validateFunc := exec.Validate()
-	if err := validateFunc(); err != nil {
-		processError(ctx, err, nil, pipelineId, cacheService, pb.Status_STATUS_VALIDATION_ERROR)
+	go validateFunc(doneChannel, errorChannel)
+
+	select {
```

Comment:
seems like code duplication (e.g. line 278) can we extract it into a function?

Reply:
Changed.